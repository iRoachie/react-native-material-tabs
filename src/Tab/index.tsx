import React from 'react'

import Entypo from 'react-native-vector-icons/Entypo'
import EvilIcons from 'react-native-vector-icons/EvilIcons'
import FontAwesome from 'react-native-vector-icons/FontAwesome'
import Foundation from 'react-native-vector-icons/Foundation'
import Ionicons from 'react-native-vector-icons/Ionicons'
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons'
import MaterialIcons from 'react-native-vector-icons/MaterialIcons'
import SimpleLineIcons from 'react-native-vector-icons/SimpleLineIcons'
import Zocial from 'react-native-vector-icons/Zocial'

import {
  TabText,
  TabBody,
  TabButton,
  IconStyle
} from './styles'


interface TabProps {
  text: string
  iconSet: string
  tabWidth: number
  stretch: boolean
  activeTextColor: string
  contentType: string
  inActiveTextColor: string
  active?: boolean
  onPress?(): void
}

const Tab: React.SFC<TabProps> = ({activeTextColor, active, onPress, text, inActiveTextColor, tabWidth, stretch, contentType, iconSet}) => {
  const color = active ? activeTextColor : inActiveTextColor

  const renderTextTab = (color, text) => {
    return (
      <TabText color={color}>{text.toUpperCase()}</TabText>
    )
  }

  const renderIconTab = (color, text, iconSet) => {
    switch (iconSet) {
      case 'Entypo':
        return (<Entypo name={text} style={[{color: color}, IconStyle]}/>)
      case 'EvilIcons':
        return (<Entypo name={text} style={[{color: color}, IconStyle]}/>)
      case 'FontAwesome':
        return (<FontAwesome name={text} style={[{color: color}, IconStyle]}/>)
      case 'Foundation':
        return (<Foundation name={text} style={[{color: color}, IconStyle]}/>)
      case 'Ionicons':
        return (<Ionicons name={text} style={[{color: color}, IconStyle]}/>)
      case 'MaterialCommunityIcons':
        return (<MaterialCommunityIcons name={text} style={[{color: color}, IconStyle]}/>)
      case 'MaterialIcons':
        return (<MaterialIcons name={text} style={[{color: color}, IconStyle]}/>)
      case 'SimpleLineIcons':
        return (<SimpleLineIcons name={text} style={[{color: color}, IconStyle]}/>)
      case 'Zocial':
        return (<Zocial name={text} style={[{color: color}, IconStyle]}/>)
      default:
        return (<Ionicons name={text} style={[{color: color}, IconStyle]}/>)
    }
  }

  const renderTab = (contentType, color, text, iconSet) => {
    switch (contentType) {
      case 'text':
        return renderTextTab(color, text)
      case 'icon':
        return renderIconTab(color, text, iconSet)
      default:
        return renderTextTab(color, text)
    }
  }

  return (
    <TabButton onPress={onPress} tabWidth={tabWidth} stretch={stretch}>
      <TabBody>
        { renderTab(contentType, color, text, iconSet) }
      </TabBody>
    </TabButton>
  )
}

export default Tab
