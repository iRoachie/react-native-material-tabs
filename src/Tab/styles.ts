import styled from 'styled-components/native'
import {Platform} from 'react-native'
import Button from '../Touchable'
import values from '../values'

export const TabBody = styled.View`
  height: ${values.barHeight}
  alignItems: center
  justifyContent: center
  paddingHorizontal: 12
`

interface TabButtonProps {
  tabWidth: number
  stretch: boolean
}

export const TabButton = styled(Button)`
  width: ${(props: TabButtonProps) => props.tabWidth}
`

interface TabTextProps {
  color: string
}

export const IconStyle = {
  fontSize: 28,
  height: 30
}

export const TabText = styled.Text`
  color: ${(props: TabTextProps) => props.color}
  fontWeight: ${Platform.OS === 'ios' ? 500 : 400}
  fontFamily: ${Platform.OS === 'android' ? 'sans-serif-medium' : 'System'}
  fontSize: 14
  textAlign: center
  minWidth: 100%
`
