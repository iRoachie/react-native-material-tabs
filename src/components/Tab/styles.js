// @flow

import styled from 'styled-components/native';
import { Platform } from 'react-native';
import Button from '../Touchable';
import type { StyleObj } from '../../lib/definitions';

type TabBodyProps = {
  tabHeight: number,
  activeTabStyle: StyleObj,
};

export const TabBody = styled.View`
  height: ${(props: TabBodyProps) => props.tabHeight};
  align-items: center;
  justify-content: center;
  padding-horizontal: 12px;
`;

type TabButtonProps = {
  tabWidth: number,
};

export const TabButton = styled(Button)`
  width: ${(props: TabButtonProps) => props.tabWidth};
`;

type TabTextProps = {
  color: string,
};

export const TabText = styled.Text`
  color: ${(props: TabTextProps) => props.color};
  font-weight: ${Platform.OS === 'ios' ? 500 : 400};
  font-family: ${Platform.OS === 'android' ? 'sans-serif-medium' : 'System'};
  font-size: 14;
  text-align: center;
  min-width: 100%;
`;
