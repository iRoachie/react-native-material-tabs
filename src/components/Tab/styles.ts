import { Platform } from 'react-native';
import styled from 'styled-components/native';

import Button from '../Touchable';

interface TabBodyProps {
  tabHeight: number;
}

export const TabBody = styled.View`
  height: ${(props: TabBodyProps) => props.tabHeight};
  align-items: center;
  justify-content: center;
  padding-left: 12px;
  padding-right: 12px;
`;

interface TabButtonProps {
  tabWidth: number;
}

export const TabButton = styled(Button)`
  width: ${(props: TabButtonProps) => props.tabWidth};
`;

interface TabTextProps {
  color: string;
}

export const TabText = styled.Text`
  color: ${(props: TabTextProps) => props.color};
  font-weight: ${Platform.OS === 'ios' ? 500 : 400};
  font-family: ${Platform.OS === 'android' ? 'sans-serif-medium' : 'System'};
  font-size: 14;
  text-align: center;
  min-width: 100%;
`;
