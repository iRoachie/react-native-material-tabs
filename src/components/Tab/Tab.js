// @flow

import React from 'react';
import { StyleSheet } from 'react-native';
import { TabText, TabBody, TabButton } from './styles';
import type { StyleObj } from '../../lib/definitions';

type TabProps = {
  allowFontScaling: boolean,
  text: string,
  tabWidth: number,
  tabHeight: number,
  stretch: boolean,
  activeTextColor: string,
  inActiveTextColor: string,
  active?: boolean,
  textStyle: StyleObj,
  uppercase: boolean,
  activeTextStyle?: StyleObj,
  onPress?: () => void,
};

const Tab = ({
  allowFontScaling,
  activeTextColor,
  active,
  onPress,
  text,
  inActiveTextColor,
  tabWidth,
  tabHeight,
  stretch,
  textStyle,
  uppercase,
  activeTextStyle,
}: TabProps) => {
  const color = active ? activeTextColor : inActiveTextColor;

  return (
    <TabButton onPress={onPress} tabWidth={tabWidth} stretch={stretch}>
      <TabBody tabHeight={tabHeight}>
        <TabText
          color={color}
          style={StyleSheet.flatten([textStyle, activeTextStyle])}
          allowFontScaling={allowFontScaling}
        >
          {(typeof(text) === "string" && uppercase) ? text.toUpperCase() : text}
        </TabText>
      </TabBody>
    </TabButton>
  );
};

export default Tab;
