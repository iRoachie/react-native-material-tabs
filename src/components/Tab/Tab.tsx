import React from 'react';
import { StyleSheet, StyleProp, TextStyle } from 'react-native';
import { TabText, TabBody, TabButton } from './styles';

export type ContentType = string | React.ReactElement;

interface TabProps {
  allowFontScaling: boolean;
  content: ContentType;
  tabWidth: number;
  tabHeight: number;
  activeTextColor: string;
  inActiveTextColor: string;
  active?: boolean;
  textStyle: StyleProp<TextStyle>;
  uppercase: boolean;
  activeTextStyle?: StyleProp<TextStyle>;
  onPress(): void;
}

const Tab = ({
  allowFontScaling,
  activeTextColor,
  active,
  onPress,
  content,
  inActiveTextColor,
  tabWidth,
  tabHeight,
  textStyle,
  uppercase,
  activeTextStyle,
}: TabProps) => {
  const color = active ? activeTextColor : inActiveTextColor;

  return (
    <TabButton onPress={onPress} tabWidth={tabWidth}>
      <TabBody tabHeight={tabHeight}>
        {typeof content === 'string' ? (
          <TabText
            color={color}
            style={StyleSheet.flatten([textStyle, activeTextStyle])}
            allowFontScaling={allowFontScaling}
          >
            {uppercase ? content.toUpperCase() : content}
          </TabText>
        ) : (
          React.cloneElement(content, {
            style: [content.props.style, { color }],
          })
        )}
      </TabBody>
    </TabButton>
  );
};

Tab.defaultProps = {
  active: false,
  onPress: () => null,
};

export default Tab;
