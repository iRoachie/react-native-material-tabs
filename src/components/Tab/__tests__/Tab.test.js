import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { create } from 'react-test-renderer';
import { configure } from 'enzyme';
import { Text } from 'react-native';

import Tab from '../Tab';

configure({ adapter: new Adapter() });

describe('Components | Tab', () => {
  describe('when content is react element', () => {
    it('renders content', () => {
      const component = (
        <Tab
          allowFontScaling
          content={<Text>hi</Text>}
          tabWidth={100}
          tabHeight={100}
          activeTextColor="red"
          inActiveTextColor="green"
          textStyle={false}
          uppercase
        />
      );
      const tree = create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when content is string', () => {
    it('renders text tab', () => {
      const component = (
        <Tab
          allowFontScaling
          content="text"
          tabWidth={100}
          tabHeight={100}
          activeTextColor="red"
          inActiveTextColor="green"
          textStyle={false}
          uppercase
        />
      );
      const tree = create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
