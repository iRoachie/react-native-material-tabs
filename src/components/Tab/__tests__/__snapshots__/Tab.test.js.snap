// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components | Tab when content is react element renders content 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
      "width": 100,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "height": 100,
          "justifyContent": "center",
          "paddingLeft": 12,
          "paddingRight": 12,
        },
      ]
    }
    tabHeight={100}
  >
    <Text
      style={
        Array [
          undefined,
          Object {
            "color": "green",
          },
        ]
      }
    >
      hi
    </Text>
  </View>
</View>
`;

exports[`Components | Tab when content is string renders text tab 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
      "width": 100,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "height": 100,
          "justifyContent": "center",
          "paddingLeft": 12,
          "paddingRight": 12,
        },
      ]
    }
    tabHeight={100}
  >
    <Text
      allowFontScaling={true}
      color="green"
      style={
        Array [
          Object {
            "color": "green",
            "fontFamily": "System",
            "fontSize": 14,
            "fontWeight": "500",
            "minWidth": "100%",
            "textAlign": "center",
          },
          Object {},
        ]
      }
    >
      TEXT
    </Text>
  </View>
</View>
`;
